### PR TITLE
Live s

### PR DESCRIPTION
**Pull request by asdf23**
Adds support  for Live S

### CODE REVIEW

**Review by foxxyz — CHANGES_REQUESTED** [review]
This is awesome, @asdf23!

Would you be able to update the following as well:
 - README.md: Add new screen name as an option and indicate Live S support
 - [device.test.js](https://github.com/foxxyz/loupedeck/blob/master/test/device.test.js#L110): Can you add tests for writing to this new display, as well as a test for `getInfo()` that includes the data that the Live S sends back? This would help so I don't accidentally regress in the future as I'm unable to personally test on a Live S.

**Comment by latenitefilms** [comment]
You'll also need to re-work `drawKey` as that references the `center` display:

https://github.com/foxxyz/loupedeck/blob/7c5d99b9d659af6b6bf7c79da6fb6f43c8724693/device.js#L130

The button count is also different, and you don't have the "side" screens (so you need to allow for 15 pixels of padding on either side) - so you'll probably need to adjust things based on the `productId` of the device (which looks like it's already being captured for devices using the latest serial firmware?).

For example, I made these quick and dirty tweaks to make it draw correct using the `simple` example:

```js
    // Draw to a specific key index (0-12)
    drawKey(index, cb) {
        // Get offset x/y for key index
        const width = 90
        const height = 90
        const x = (30/2) + index % 5 * width
        const y = Math.floor(index / 5) * height
        return this[cb instanceof Buffer ? 'drawBuffer' : 'drawCanvas']({ id: 'main', x, y, width, height }, cb)
    }
```

You'll also need to update `onTouch` with similar maths. For example:

```js
onTouch(event, buff) {
        const x = buff.readUInt16BE(1)
        const y = buff.readUInt16BE(3)
        const id = buff[5]

        // Determine target
        const screen = 'main'
        let key
        if (screen === 'main') {
            const column = Math.floor(((x + 15) - 60) / 90)
            const row = Math.floor(y / 90)
            key = row * 5 + column
        }

        // Create touch
        const touch = { x, y, id, target: { screen, key } }

        // End touch, remove from local cache
        if (event === 'touchend') {
            delete this.touches[touch.id]
        } else {
            // First time seeing this touch, emit touchstart instead of touchmove
            if (!this.touches[touch.id]) event = 'touchstart'
            this.touches[touch.id] = touch
        }

        this.emit(event, { touches: Object.values(this.touches), changedTouches: [touch] })
    }
```

I'm also not sure that the bottom left button should be labelled "circle" - it probably makes more sense for the Loupedeck Live-S if the bottom left is button 1, then the top right is button 2, etc.

Again, obviously you need to tweak things so that it's a per-device thing, so you don't break Loupedeck Live support. Once this is done however, it should be pretty easy to add Loupedeck CT,  and Razer Stream Controller support.

Hope this helps!

**Comment by latenitefilms** [comment]
Also, `getInfo` seems to work fine here. Here's the output from the `simple` example with those above tweaks:

```
✅ Connected to Loupedeck at /dev/tty.usbmodem32101
Device serial number LDD2201010022300100020B1001, software version 0.1.1
Knob knobTL rotated left
Testing vibration #0
Knob knobTL rotated right
Testing vibration #1
Knob knobCL rotated left
Knob knobCL rotated right
Button circle pressed
Button circle released
Button 1 pressed
Button 1 released
Button 2 pressed
Button 2 released
Button 3 pressed
Button 3 released
Touch #21 started: x: 69, y: 47
Touch #21 ended: x: 69, y: 47
Touch #18 started: x: 160, y: 52
Touch #18 ended: x: 160, y: 52
Touch #15 started: x: 250, y: 48
Touch #15 ended: x: 250, y: 48
Touch #17 started: x: 338, y: 30
Touch #17 ended: x: 338, y: 30
Touch #17 started: x: 443, y: 41
Touch #17 ended: x: 443, y: 41
Touch #20 started: x: 70, y: 129
Touch #20 ended: x: 70, y: 129
Touch #19 started: x: 168, y: 120
Touch #19 ended: x: 168, y: 120
Touch #20 started: x: 258, y: 131
Touch #20 ended: x: 258, y: 131
Touch #20 started: x: 331, y: 129
Touch #20 ended: x: 331, y: 129
Touch #15 started: x: 429, y: 129
Touch #15 ended: x: 429, y: 129
Touch #19 started: x: 69, y: 221
Touch #19 ended: x: 69, y: 221
Touch #20 started: x: 147, y: 210
Touch #20 ended: x: 147, y: 210
Touch #19 started: x: 253, y: 225
Touch #19 ended: x: 253, y: 225
Touch #18 started: x: 344, y: 221
Touch #18 ended: x: 344, y: 221
Touch #16 started: x: 440, y: 223
Touch #16 ended: x: 440, y: 223
```

**Comment by foxxyz** [comment]
Thank you @latenitefilms! This is really useful, and it looks this PR likely requires a bit more work.

If I had access to a device I'd definitely add all the support for it. 

Are you able to confirm what firmware version is running on the device?

**Comment by asdf23** [comment]
Sorry I've never done a pull request on github, I didnt' see these suggestions until just now.  I do have an update that I think addresses these issues.

File: constants.js [hunk 1]
```diff
@@ -24,6 +24,7 @@ const DISPLAYS = {
     center: { id: Buffer.from('\x00A'), width: 360, height: 270 }, // "A"
     left: { id: Buffer.from('\x00L'), width: 60, height: 270 }, // "L"
     right: { id: Buffer.from('\x00R'), width: 60, height: 270 }, // "R"
+    live_s: { id: Buffer.from('\x00M'), width: 480, height: 270 }, // "M"
```
[the remark this answers is not in the document]
Comment (asdf23) [reply constants.js]: Sorry I didn't see this comment, and already pushed another change. I have not found a way to automatically detect which device is plugged in. The touch events need to know which device so I have a work around (in a later commit)

File: device.js [hunk 2]
```diff
@@ -24,6 +24,49 @@ class LoupedeckDevice extends EventEmitter {
 
         return rawDevices.flat()
     }
+    static thisDevice = "center";
+    setDevice(x) {
+        this.thisDevice = x;
```
Comment (asdf23) [inline device.js:29]: Work around. Ideally this would detect the device automatically

Comment (foxxyz) [reply device.js:29]: Not sure I understand what this is for?

File: device.js [hunk 3]
```diff
@@ -24,6 +24,44 @@ class LoupedeckDevice extends EventEmitter {
...
+                let topBorder = bottomBorder + DISPLAYS[screen].buttonHeight;
+                if((x>= leftBorder && x < rightBorder) && (y>=bottomBorder && y<topBorder)) {
+                    return buttonIndex;
+                }
+                buttonIndex++;
+            }
+        }
+        return -1;
+    }
+    buttonIndexToKey(buttonIndex) {
```
Comment (foxxyz) [inline device.js:57]: this could benefit from a variable destructure for readability

File: device.js [hunk 4]
```diff
@@ -24,6 +24,44 @@ class LoupedeckDevice extends EventEmitter {
...
+    getScreen(x,y) {
+        switch(this.thisDevice) {
+            default:
+            case "center":
+                return WHICH_DISPLAY.find(wd=> x <= wd.barrier).screen;
+            case "main":
+                return this.thisDevice;
+        }
+    }
+    getButton(x,y) {
```
Comment (foxxyz) [inline device.js:40]: please add some tests to make sure this method works correctly.

I disagree on returning the value `-1`. It should probably `throw` if the x/y coordinates are invalid for any given reason.

**Review by foxxyz — CHANGES_REQUESTED** [review]
Please make sure all the existing tests continue to function (run `npm test` locally) and add new tests for the functionality you're adding.

**Comment by foxxyz** [comment]
Closing this in favor of the simpler and tested implementation in #21. 

Thank you for all your hard work @asdf23!

**Pull request closed without merging**